### ISSUE TITLE
Dynamically identify filesystems supported by TensorFlow

### ISSUE THREAD

**Issue by peteboothroyd** [tensorflow/tfx#2877]
Not a Contribution.

TFX has a filesystem concept which can leverage TensorFlow for interacting with a range of different filesystems. The schemas that will leverage TensorFlow are currently hardcoded in the [TensorflowFilesystem](https://github.com/tensorflow/tfx/blob/c1c2503933243fe51bb14f7d63222f958178d7e7/tfx/dsl/io/plugins/tensorflow_gfile.py#L32) class. However TensorFlow supports [pluggable filesystems](https://github.com/tensorflow/community/blob/master/rfcs/20190506-filesystem-plugin-modular-tensorflow.md), meaning that this may not reflect the full TensorFlow capabilities at runtime.

It would be great if TFX could instead dynamically detect the schemas that TF can handle, and delegate appropriately at runtime. There is a cpp method which exposes the supported schemas [FileSystemRegistryImpl::GetRegisteredFileSystemSchemes](https://github.com/tensorflow/tensorflow/blob/dc3099c444d294b39cd79fe1d1a4bff59a0c6180/tensorflow/core/platform/env.cc#L101), which would need to be exposed through the python api, but would meet the needs of this feature.

There is another complication in that the TFX filesystem class currently registers schemas recognised by TensorFlow at [import time](https://github.com/tensorflow/tfx/blob/fcb9548df7445afb4bad2cab4e1837aebe7556f4/tfx/dsl/io/plugins/tensorflow_gfile.py#L91), meaning that if a filesystem plugin is loaded after TFX, then TFX would be unaware that TensorFlow can now support that filesystem.

Note I believe that a workaround for now is to place this before trying to use the new filesystem schema:

```python
from tfx.dsl.io.plugins import tensorflow_gfile
from tfx.dsl.io import filesystem_registry
tensorflow_gfile.TensorflowFilesystem.SUPPORTED_SCHEMES.append("X://")
filesystem_registry.DEFAULT_FILESYSTEM_REGISTRY.register(tensorflow_gfile.TensorflowFilesystem, priority=0)
```

**Comment by zhitaoli** [comment]
@charlesccychen Can you look at this since you are touching filesystem part recently?

@peteboothroyd : do you a particular filesystem which cannot be utilized by TFX but should be addressed by this? Or blocker to something other capability?

>  pluggable filesystems,

I'm not sure that proposal has been implemented or not yet?

>  There is a cpp method which exposes the supported schemas FileSystemRegistryImpl::GetRegisteredFileSystemSchemes, which would need to be exposed through the python api, but would meet the needs of this feature.

This might be a problem, as we have not directly touched the cpp level so far. If TensorFlow exposes some knobs in its python API we could certainly look at utilizing it.

**Comment by peteboothroyd** [comment]
Thanks for getting back so quickly!

> do you a particular filesystem which cannot be utilized by TFX but should be addressed by this? Or blocker to something other capability?

I think the workaround I posted unblocks me. It seems somewhat hacky to be reaching in and modifying the global TFX state and then reregistering the implementation though.

> I'm not sure that proposal has been implemented or not yet?

The linked proposal is just for a stable ABI, there has been support for a while through the [cpp](https://github.com/tensorflow/tensorflow/blob/516ae286f6cc796e646d14671d94959b129130a4/tensorflow/core/platform/file_system.h#L861-L864) interface.

**Comment by charlesccychen** [comment]
@peteboothroyd We have merged #2953 which should address this issue.  Let me know if you have any additional comments or suggestions.  Thanks!

**Comment by peteboothroyd** [comment]
that's great, thank you for the help!
